unit_tests/src/test_rc_update_replica_access_time.cpp

```diff
@@ -193,6 +241,114 @@ TEST_CASE("#8260: rc_update_replica_access_time can update access time of multip
+                }
+            }}
+        }.dump();
+        // clang-format on
+
+        // Show that invoking the API does not result in an error.
+        char* ignored{};
+        CHECK(rc_update_replica_access_time(conn_ptr, updates.c_str(), &ignored) == 0);
+
+        // Show that no data object exists matching the non-existent data id. This helps
+        // prove the post-condition state is what we expect. That is, no new replica is
+        // accidentally created.
+        {
+            auto query = irods::query{conn_ptr, fmt::format("select DATA_ID where DATA_ID = '{}'", non_existent_data_id)};
+            CHECK(query.empty());
+        }
+    }
+
+    SECTION("targeting existent and non-existent replica")
+    {
+        // Create a data object.
+        const auto logical_path = sandbox / "data_object.txt";
+        io::client::native_transport tp{conn};
+        io::odstream{tp, logical_path};
+
+        // Show that the replica's atime and mtime are identical.
+        const auto [data_id, atime, mtime] = get_replica_info(conn, logical_path, 0);
+        REQUIRE(atime == mtime);
+
+        // This JSON string contains the new atime for the replicas. Only the update
+        // involving the existent replica should succeed. The other update is effectively
+        // a no-op.
+        const auto* expected_atime = "01700000000";
+        // clang-format off
+        const auto updates = json{
+            {"access_time_updates", json::array_t{
+                {
+                    {"data_id", std::stoull(data_id)},
+                    {"replica_number", 0},
+                    {"atime", expected_atime}
+                },
+                // The non-existent replica.
+                {
+                    {"data_id", non_existent_data_id},
+                    {"replica_number", non_existent_replica_number},
+                    {"atime", "01600000000"}
+                }
+            }}
+        }.dump();
+        // clang-format on
+
+        // Show that invoking the API does not result in an error.
+        char* ignored{};
+        CHECK(rc_update_replica_access_time(conn_ptr, updates.c_str(), &ignored) == 0);
+
+        // Show that no data object exists matching the non-existent data id. This helps
+        // prove the post-condition state is what we expect. That is, no new replica is
+        // accidentally created.
+        {
+            auto query = irods::query{conn_ptr, fmt::format("select DATA_ID where DATA_ID = '{}'", non_existent_data_id)};
+            CHECK(query.empty());
+        }
+
+        // Show that the replica's atime has been updated to the expected value.
+        const auto updated_replica_info = get_replica_info(conn, logical_path, 0);
+        CHECK(std::get<0>(updated_replica_info) == data_id);
+        CHECK(std::get<1>(updated_replica_info) == expected_atime); // atime should be different.
+        CHECK(std::get<2>(updated_replica_info) == mtime); // mtime should not have changed.
```

Comment:
I was not expecting this result. So the inputs are vetted before attempting to perform the atomic update? Is it possible for the input to be sanitized, one replica from the input is removed (leaving at least one valid update), and then the atomic update fails? Would we then see an error, or is there still partial success?

Note: I'm not suggesting we make a test for this because it seems like that's going to be on the order of sub-second timing which we don't need any of in our tests :)

Reply:
Noting here that we discussed this and agreed that this comment can be marked as resolved.

The API does not check the values being inserted into the table are valid timestamps. The API can only fail if the inputs violate database table constraints. Partial success is possible and allowed because the API does not require the data id and replica number to match anything. The lack of a match just means there was nothing to update.

Also, this API requires rodsadmin privileges.